src/handler/mod.rs

```diff
@@ -728,13 +728,14 @@ impl Handler {
         self.new_session(node_address, session);
     }
 
-    /// Verifies a Node ENR to it's observed address. If it fails, any associated session is also
-    /// considered failed. If it succeeds, we notify the application.
+    /// Verifies a Node ENR to its observed node ID
     fn verify_enr(&self, enr: &Enr, node_address: &NodeAddress) -> bool {
-        // If the ENR does not match the observed IP addresses, we consider the Session
-        // failed.
+        // Peers will often connect from one endpoint and broadcast an ENR from another.
+        // For example, you might connect to a peer from behind a NAT, but their ENR is for their
+        // public IP address. So we accept mismatched IP & port.
+        // TODO distinguish between peers with matched & mismatched IPs and treat them differently,
+        // see TrustedState that was removed by commit 2447c3c45b4d2116ee9e8919c51f4f059587cc19
```

Comment:
I've spent a bit of time in this kind of logic, and I'm curious about the use-case you need here. 

I think your comment may be around the wrong way here. 

It is quite common for nodes to connect from behind a NAT and typically their ENR will represent their local IP/PORT which is not contactable. If their ENR is for their public IP address, then it is contactable and this check passes. 

The reason we filter these ENRs is we want to clean up the local routing table as much as possible. There is no point adding ENRs with for example Ips of `127.0.0.1`. This is non-contactable. If we add this to our routing table, then we try and attempt to send packets to these addresses in order to ping them and request results in queries. There are a number of downside to doing this:
1. Adding non-contactable ENRs to our routing table increase bandwidth and slow down queries. If we do a query and half the peers in our routing table are non-contactable, we will not get a response for all the peers we try to contact in our routing table and have to wait the request timeout duration to realise the peer is not contactable. We also then mark the peer as disconnected and eventually remove it, defeating the purpose of removing this check here. 
2. I've been very careful in requests, to only responsd to the IP's that contacted us. If we allow anyone to send us an ENR with an arbitrary IP/Port, they can force the local node to send packets to arbitrary destinations. This in principle could be used to get lots of nodes to DoS a victim by sending lots of discv5 packets, also we run into issues when we try and contact ourselves when we permit ENR's with localhost as the IP and this raises even more issues. 

I dont understand the reasoning for why you would want these nodes registered in the local routing table, but there is perhaps a use-case I'm missing.

Reply:
There is probably a long conversation to have on this. But the way it was designed is the following:
If a node starts up and doesn't know its external IP (assume this is probably the default for most nodes), it should set its ENR-IP and PORT to `None`.  If a node is specifically advertising the wrong IP and or is behind a NAT that makes it non-contactable, it has no value to us, because we can't contact it. 

Once a node connects to a set number of peers, discv5 (via PONG responses) will update its ENR to the correct external IP and port address that makes it contactable. Once this update occurs, the ENR IP and PORT will be its public versions and be a valid/useful node that we add to our routing table.

Reply:
Going to loop in @carver to discuss more, but I think I understand your points. I _think_ we probably need to take the approach of setting ENR-IP/PORT to `None`.

Reply:
Yeah. I think its important that we allow peers behind a NAT to make requests, but we need to be careful about peers that are falsely advertising their addresses so that we don't try and contact those addresses. 

The current solution is that we allow requests from peers with `None` udp sockets but reject requests from peers with invalid/incorrect ENR's. We allow the requests from `None` but we don't add them to the routing table as we have no address to contact them making sure the collected peers are all useful. See here: https://github.com/sigp/discv5/blob/master/src/service.rs#L1172

Reply:
I've removed the ENR changes for the time being.